File: references/repos/novelforge/backend/app/api/endpoints/relation_graph.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session

from app.db.session import get_session
from app.schemas.relation_graph import (
    RelationGraphBatchAppendEventsRequest,
    RelationGraphBatchCreateRequest,
    RelationGraphBatchDeleteRequest,
    RelationGraphBatchUpdateKindRequest,
    RelationGraphBatchUpdateStanceRequest,
    RelationGraphDeleteRequest,
    RelationGraphExportRequest,
    RelationGraphExportResponse,
    RelationGraphImportRequest,
    RelationGraphImportResponse,
    RelationGraphListRequest,
    RelationGraphListResponse,
    RelationGraphMetaResponse,
    RelationGraphRecord,
    RelationGraphUpsertRequest,
    RelationGraphWriteResponse,
)
from app.services.relation_graph_service import RelationGraphService


router = APIRouter()


def _service(session: Session) -> RelationGraphService:
    return RelationGraphService(session)


@router.get("/meta", response_model=RelationGraphMetaResponse, summary="关系类型与立场元数据")
def get_meta(session: Session = Depends(get_session)):
    return _service(session).get_meta()


@router.post("/list", response_model=RelationGraphListResponse, summary="分页查询关系图")
def list_relations(req: RelationGraphListRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).list_relations(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/upsert", response_model=RelationGraphRecord, summary="新增或更新关系")
def upsert_relation(req: RelationGraphUpsertRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).upsert_relation(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/delete", response_model=RelationGraphWriteResponse, summary="删除单条关系")
def delete_relation(req: RelationGraphDeleteRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).delete_relation(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/batch/delete", response_model=RelationGraphWriteResponse, summary="批量删除关系")
def batch_delete(req: RelationGraphBatchDeleteRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).batch_delete_relations(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/batch/update-kind", response_model=RelationGraphWriteResponse, summary="批量修改关系类型")
def batch_update_kind(req: RelationGraphBatchUpdateKindRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).batch_update_kind(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/batch/update-stance", response_model=RelationGraphWriteResponse, summary="批量修改立场")
def batch_update_stance(req: RelationGraphBatchUpdateStanceRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).batch_update_stance(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/batch/append-events", response_model=RelationGraphWriteResponse, summary="批量追加事件摘要")
def batch_append_events(req: RelationGraphBatchAppendEventsRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).batch_append_events(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/batch/create", response_model=RelationGraphWriteResponse, summary="批量新增关系（冲突覆盖）")
def batch_create(req: RelationGraphBatchCreateRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).batch_create_relations(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/export", response_model=RelationGraphExportResponse, summary="导出关系图数据")
def export_relations(req: RelationGraphExportRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).export_relations(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/import", response_model=RelationGraphImportResponse, summary="导入关系图数据")
def import_relations(req: RelationGraphImportRequest, session: Session = Depends(get_session)):
    try:
        return _service(session).import_relations(req)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

File: references/repos/novelforge/frontend/src/renderer/src/api/relationGraph.ts
import http from './request'
import type { components } from '@renderer/types/generated'

export type RelationGraphKey = components['schemas']['RelationGraphKey']
export type RelationGraphEvent = components['schemas']['RelationGraphEvent']
export type RelationGraphInput = components['schemas']['RelationGraphInput']
export type RelationGraphRecord = components['schemas']['RelationGraphRecord']

export type RelationGraphListRequest = components['schemas']['RelationGraphListRequest']
export type RelationGraphListResponse = components['schemas']['RelationGraphListResponse']

export type RelationGraphUpsertRequest = components['schemas']['RelationGraphUpsertRequest']
export type RelationGraphDeleteRequest = components['schemas']['RelationGraphDeleteRequest']
export type RelationGraphWriteResponse = components['schemas']['RelationGraphWriteResponse']

export type RelationGraphBatchDeleteRequest = components['schemas']['RelationGraphBatchDeleteRequest']
export type RelationGraphBatchUpdateKindRequest = components['schemas']['RelationGraphBatchUpdateKindRequest']
export type RelationGraphBatchUpdateStanceRequest = components['schemas']['RelationGraphBatchUpdateStanceRequest']
export type RelationGraphBatchAppendEventsRequest = components['schemas']['RelationGraphBatchAppendEventsRequest']
export type RelationGraphBatchCreateRequest = components['schemas']['RelationGraphBatchCreateRequest']

export type RelationGraphExportRequest = components['schemas']['RelationGraphExportRequest']
export type RelationGraphExportResponse = components['schemas']['RelationGraphExportResponse']
export type RelationGraphImportRequest = components['schemas']['RelationGraphImportRequest']
export type RelationGraphImportResponse = components['schemas']['RelationGraphImportResponse']
export type RelationGraphKindOption = components['schemas']['RelationGraphKindOption']
export type RelationGraphMetaResponse = components['schemas']['RelationGraphMetaResponse']

export type RelationGraphKind = NonNullable<components['schemas']['RelationGraphInput']['kind_cn']>
export type RelationGraphStance = NonNullable<components['schemas']['RelationGraphInput']['stance']>

export function getRelationGraphMeta() {
  return http.get<RelationGraphMetaResponse>('/relation-graph/meta')
}

export function listRelationGraph(data: RelationGraphListRequest) {
  return http.post<RelationGraphListResponse>('/relation-graph/list', data)
}

export function upsertRelationGraph(data: RelationGraphUpsertRequest) {
  return http.post<RelationGraphRecord>('/relation-graph/upsert', data)
}

export function deleteRelationGraph(data: RelationGraphDeleteRequest) {
  return http.post<RelationGraphWriteResponse>('/relation-graph/delete', data)
}

export function batchDeleteRelationGraph(data: RelationGraphBatchDeleteRequest) {
  return http.post<RelationGraphWriteResponse>('/relation-graph/batch/delete', data)
}

export function batchUpdateKindRelationGraph(data: RelationGraphBatchUpdateKindRequest) {
  return http.post<RelationGraphWriteResponse>('/relation-graph/batch/update-kind', data)
}

export function batchUpdateStanceRelationGraph(data: RelationGraphBatchUpdateStanceRequest) {
  return http.post<RelationGraphWriteResponse>('/relation-graph/batch/update-stance', data)
}

export function batchAppendEventsRelationGraph(data: RelationGraphBatchAppendEventsRequest) {
  return http.post<RelationGraphWriteResponse>('/relation-graph/batch/append-events', data)
}

export function batchCreateRelationGraph(data: RelationGraphBatchCreateRequest) {
  return http.post<RelationGraphWriteResponse>('/relation-graph/batch/create', data)
}

export function exportRelationGraph(data: RelationGraphExportRequest) {
  return http.post<RelationGraphExportResponse>('/relation-graph/export', data)
}

export function importRelationGraph(data: RelationGraphImportRequest) {
  return http.post<RelationGraphImportResponse>('/relation-graph/import', data)
}
